File: frontend/src/types/geospatial.ts
export interface StateSpatialItem {
  state: string;
  state_code: number;
  lat: number;
  lon: number;
  region: string;
  agro_zone: string;
  district_count: number;
  average_yield_kg_ha: number;
  average_area_k_ha: number;
  risk_score: number;
  risk_level: 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL';
  trend_direction: string;
  theil_sen_slope: number;
  forecast_1yr_kg_ha: number;
  anomaly_count: number;
  cluster_id: number;
}

export interface GeospatialOverviewResponse {
  total_states_monitored: number;
  total_districts_monitored: number;
  total_observations: number;
  national_average_yield_kg_ha: number;
  national_average_risk_score: number;
  high_risk_states_count: number;
  anomalous_states_count: number;
  spatial_outliers_count: number;
  clusters_count: number;
}

export interface SpatialClusterItem {
  cluster_id: number;
  cluster_name: string;
  archetype: string;
  risk_profile: string;
  district_count: number;
  avg_yield_kg_ha: number;
  avg_volatility_pct: number;
  avg_theil_sen_slope: number;
  avg_anomaly_rate_pct: number;
  dominant_states: Record<string, number>;
}

export interface SpatialOutlierItem {
  state: string;
  district: string;
  yield_kg_ha: number;
  state_mean_yield: number;
  within_state_zscore: number;
  relative_yield_ratio: number;
  relative_risk_ratio: number;
  anomaly_rate_pct: number;
  theil_sen_slope: number;
  reasons: string[];
  severity: 'MODERATE' | 'HIGH';
}

export interface StateSpatialProfileResponse {
  state: string;
  state_code: number;
  lat: number;
  lon: number;
  region: string;
  agro_zone: string;
  district_count: number;
  average_yield_kg_ha: number;
  risk_score: number;
  risk_level: string;
  trend_direction: string;
  theil_sen_slope: number;
  forecasts: Array<{
    forecast_year: number;
    horizon_years: number;
    predicted_yield: number;
    lower_bound_p10: number;
    upper_bound_p90: number;
    uncertainty_pct: number;
  }>;
  spatial_outliers: SpatialOutlierItem[];
  districts: Array<{
    district_name: string;
    average_yield_kg_ha: number;
    yield_volatility_pct: number;
    state_relative_yield_ratio: number;
    district_yield_zscore_state: number;
    trend_direction: string;
    trend_theil_sen_slope: number;
    anomaly_rate_pct: number;
  }>;
}

export interface SimilarRegionItem {
  state: string;
  district: string;
  similarity_score: number;
  average_yield_kg_ha: number;
  yield_volatility_pct: number;
  trend_theil_sen_slope: number;
}

File: backend/schemas/geospatial.py
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field

class StateSpatialItem(BaseModel):
    state: str
    state_code: int
    lat: float
    lon: float
    region: str
    agro_zone: str
    district_count: int
    average_yield_kg_ha: float
    average_area_k_ha: float
    risk_score: float
    risk_level: str
    trend_direction: str
    theil_sen_slope: float
    forecast_1yr_kg_ha: float
    anomaly_count: int
    cluster_id: int

class GeospatialOverviewResponse(BaseModel):
    total_states_monitored: int
    total_districts_monitored: int
    total_observations: int
    national_average_yield_kg_ha: float
    national_average_risk_score: float
    high_risk_states_count: int
    anomalous_states_count: int
    spatial_outliers_count: int
    clusters_count: int

class SpatialClusterItem(BaseModel):
    cluster_id: int
    cluster_name: str
    archetype: str
    risk_profile: str
    district_count: int
    avg_yield_kg_ha: float
    avg_volatility_pct: float
    avg_theil_sen_slope: float
    avg_anomaly_rate_pct: float
    dominant_states: Dict[str, int]

class SpatialQueryRequest(BaseModel):
    state: Optional[str] = None
    district: Optional[str] = None
    metric: Optional[str] = "yield"
    risk_threshold: Optional[float] = None
    anomaly_only: Optional[bool] = False
    cluster_id: Optional[int] = None

class SimilarRegionItem(BaseModel):
    state: str
    district: str
    similarity_score: float
    average_yield_kg_ha: float
    yield_volatility_pct: float
    trend_theil_sen_slope: float
